### ISSUE TITLE
I2C Communication Timing

### ISSUE THREAD

**Issue by smartynenko** [aj-foster/FTC-Examples#4]
First of all, many thanks for awesome I2C driver and examples, it is a huge contribution to FTC community. I have a question about performance. In one of your presentations [Custom LEDs for FTC](http://ftc.flfirst.org/images/Home_Files/Documents/2019/Custom_LEDs_for_FTC.pdfl) you mentioned that it takes about 5ms to send data to I2C device. How many LEDs did you use for testing? I currently have 60 LEDs in my set up and it takes about 40ms to send data to the SPI bridge.

Thank you!

**Comment by aj-foster** [comment]
Hello there,

Thanks for the question. Sorry I didn't notice it earlier.

There are a few factors that could affect performance. To be honest, I don't remember the conditions under which we originally measured 5ms refresh rates. Using the hardware at my disposal, there does seem to be a notable relationship between length and refresh rate:

| Length | Rate (ms) |
|---:|---:|
| 1 | 8.5 |
| 5 | 10.4 |
| 10 | 11.2 |
| 20 | 17.9 |
| 40 | 35.9 |
| 60 | 50.0 |

These are incredibly rough estimates. My setup uses a REV Control Hub and the rainbow pattern included in this repository, which leads us into some of the things to consider:

- If the slowest part of the hardware is the CPU, you might get different performance depending on whether you're on a phone with a fast, multi-core CPU.
- Along the same lines, it's worth making sure there's nothing else taking up CPU cycles (other apps that need to be closed, etc.)
- If the I2C bus speed is the constraint, it's worth making sure there's nothing else taking up I2C time. Don't use I2C bus 0 on a REV hub, for example, since that's where the IMU lives.
- It's also possible that hardware isn't the problem. Maybe there are expensive calculations in between calls to `update` the LEDs.

The example rainbow pattern in this repo still looks "smooth" despite crossing the 16ms (or 60Hz) boundary because the change in color between updates is pretty narrow. I can see where other patterns would start to suffer, though.

tl;dr Your 40ms refresh rate sounds about right. Make sure there aren't other things slowing it down (calculations, other apps, other I2C devices).

**Comment by smartynenko** [comment]
Thank you for the detailed response. I did some additional testing and it looks like I2C bandwidth is a bottleneck and update speed is roughly proportional to time required to transfer all bytes in LED array, time required to populate LEDs is negligible comparing to time spent in [write](https://github.com/aj-foster/FTC-Examples/blob/a7a245619101b5d7e8048b34ca22972832d741db/LEDs/DotStar/DotStarBridgedLED.java#L343) call
